crml/staking/src/lib.rs

```diff
@@ -1688,59 +1599,38 @@ impl<T: Trait> Module<T> {
 		if let Some(controller) = <Bonded<T>>::take(stash) {
 			<Ledger<T>>::remove(&controller);
 		}
-		<Payee<T>>::remove(stash);
+		T::Rewarder::remove_payee(stash);
 		<Validators<T>>::remove(stash);
 		<Nominators<T>>::remove(stash);
 
 		slashing::clear_stash_metadata::<T>(stash);
 	}
 
-	/// Add reward points to validators using their stash account ID.
-	///
-	/// Validators are keyed by stash account ID and must be in the current elected set.
-	///
-	/// For each element in the iterator the given number of points in u32 is added to the
-	/// validator, thus duplicates are handled.
-	///
-	/// At the end of the era each the total payout will be distributed among validator
-	/// relatively to their points.
-	///
-	/// COMPLEXITY: Complexity is `number_of_validator_to_reward x current_elected_len`.
-	/// If you need to reward lots of validator consider using `reward_by_indices`.
-	pub fn reward_by_ids(validators_points: impl IntoIterator<Item = (T::AccountId, u32)>) {
-		CurrentEraPointsEarned::mutate(|rewards| {
-			let current_elected = <Module<T>>::current_elected();
-			for (validator, points) in validators_points.into_iter() {
-				if let Some(index) = current_elected.iter().position(|elected| *elected == validator) {
-					rewards.add_points_to_index(index as u32, points);
-				}
-			}
-		});
-	}
-
-	/// Add reward points to validators using their validator index.
-	///
-	/// For each element in the iterator the given number of points in u32 is added to the
-	/// validator, thus duplicates are handled.
-	pub fn reward_by_indices(validators_points: impl IntoIterator<Item = (u32, u32)>) {
-		let current_elected_len = <Module<T>>::current_elected().len() as u32;
-
-		CurrentEraPointsEarned::mutate(|rewards| {
-			for (validator_index, points) in validators_points.into_iter() {
-				if validator_index < current_elected_len {
-					rewards.add_points_to_index(validator_index, points);
-				}
-			}
-		});
-	}
-
 	/// Ensures that at the end of the current session there will be a new era.
 	fn ensure_new_era() {
 		match ForceEra::get() {
 			Forcing::ForceAlways | Forcing::ForceNew => (),
 			_ => ForceEra::put(Forcing::ForceNew),
 		}
 	}
+
+	/// Return the reward destination for a given stash id.
+	///
+	/// This is mostly a helper function for test purposes. There is a simpler interface,
+	/// HandlePayee, implemented by the rewards module which is suggested to look at first before
+	/// using this one in your code.   
+	pub fn payee(stash: &T::AccountId) -> RewardDestination<T::AccountId> {
+		let payee = T::Rewarder::payee(stash);
+		if payee == *stash {
+			return RewardDestination::Stash;
+		}
+		if let Some(controller) = Bonded::<T>::get(stash) {
+			if payee == controller {
+				return RewardDestination::Controller;
+			}
+		}
+		RewardDestination::Account(payee)
```

Comment:
this is the existing behaviour and should remain, controller always needs 2 look ups so we phased it out already on v37 by disallowing it from being created in the first place.
```suggestion
		let payee = T::Rewarder::payee(stash);
		if payee == *stash {
			return RewardDestination::Stash;
		}
		RewardDestination::Account(payee)
```

Reply:
This function is only used for tests. There's no need to optimise it. RewardDestination::Controller is a better reflection of the RewardDestination enum and says more than RewardDestination::Account().

Reply:
it's not optimise, it tests something which doesn't happen in reality now

Reply:
Even RewardDestination::Account(payee) doesn't happen in reality since we started to handle payees in the rewards module. It's just an interface that we have now made it limited to tests. The staking module gets the reward destination through the RewardDestination enum and it's nice to give it back in the same way.

Reply:
Ok RewardDestination is just a legacy thing for us at this point. We accept it as an input to some staking calls but it is never stored. The tests should assert on the stored AccountId directly